File: 416-partition-equal-subset-sum/partition-equal-subset-sum.py
class Solution:
    def canSubsets(self,nums,remender,idx,dp):
        if(remender == 0) :return True
        if(idx >= len(nums) or remender < 0): return False
        if(dp[remender][idx] != -1): return dp[remender][idx]
        take = self.canSubsets(nums,remender - nums[idx],idx+1,dp)
        skip = self.canSubsets(nums,remender,idx+1,dp)
        dp[remender][idx] = take or skip
        return dp[remender][idx]
    def canPartition(self, nums: List[int]) -> bool:
        sum1 = sum(nums)
        if(sum1%2 == 1): return False
        target = sum1//2
        dp = [[-1]* len(nums) for _ in range(target+1)]
        return self.canSubsets(nums,target,0,dp)
        

File: DP/416-partition-equal-subset-sum/partition-equal-subset-sum.java
class Solution {
    Boolean [][] dp;
    public boolean canPartition(int[] nums) {
        int sum =0;
        for(int num:nums) sum += num;
        if(sum%2 == 1) return false;
        dp = new Boolean[1+ sum/2][nums.length+1];
        return canSubsets(nums,sum/2,0);
    }
    boolean canSubsets(int[] nums,int remain,int idx){
        if(remain == 0) return true;
        if(idx >= nums.length || remain < 0) return false;
        if(dp[remain][idx] != null) return false;
        boolean take = canSubsets(nums,remain - nums[idx],idx+1);
        boolean skip = canSubsets(nums,remain,idx+1);
        return dp[remain][idx]= take || skip ;
    } 
}